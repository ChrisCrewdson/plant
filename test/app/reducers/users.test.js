const users = require('../../../app/reducers/users');
const actions = require('../../../app/actions');
const assert = require('assert');
const Immutable = require('immutable');

function checkReducer(actionName, state, payload, expected) {
  const action = actions[actionName](payload);
  const actual = users(state, action);
  // The following line provides useful debug info which the one after does not
  assert.deepEqual(actual.toJS(), expected.toJS());
  assert(Immutable.is(actual, expected));
}

describe('/app/reducers/users', () => {
  test('should reduce loadUserSuccess action', () => {
    const state = Immutable.fromJS({});
    const payload = { _id: '1', name: 'john' };
    const expected = Immutable.fromJS({
      1: { _id: '1', name: 'john', locationIds: Immutable.Set() },
    });
    checkReducer('loadUserSuccess', state, payload, expected);
  });

  test('should reduce loadUsersSuccess action', () => {
    const state = Immutable.fromJS({});
    const payload = [{ _id: '1', name: 'john' }];
    const expected = Immutable.fromJS({
      1: { _id: '1', name: 'john', locationIds: Immutable.Set() },
    });
    checkReducer('loadUsersSuccess', state, payload, expected);
  });

  test('should reduce createPlantRequest action', () => {
    const state = Immutable.fromJS({ u1: { _id: 'u1', name: 'john', locationIds: Immutable.Set(['p1']) } });
    const payload = { _id: 'p2', title: 'pt', userId: 'u1' };
    const expected = Immutable.fromJS({
      u1: { _id: 'u1', name: 'john', locationIds: Immutable.Set(['p1']) },
    });
    checkReducer('createPlantRequest', state, payload, expected);
  });

  test('should reduce loadPlantsSuccess action', () => {
    const state = Immutable.fromJS({
      u1: {
        _id: 'u1',
        name: 'john',
        locationIds: Immutable.Set(['p1.1']),
      },
      u2: {
        _id: 'u2',
        name: 'jane',
        locationIds: Immutable.Set(['p2.1', 'p2.2']),
      },
    });
    const payload = [
      { _id: 'p1.1', userId: 'u1' },
      { _id: 'p1.2', userId: 'u1' },
      { _id: 'p2.2', userId: 'u2' },
      { _id: 'p2.3', userId: 'u2' },
      { _id: 'p3.1', userId: 'u3' },
    ];
    const expected = Immutable.fromJS({
      u1: { _id: 'u1', name: 'john', locationIds: Immutable.Set(['p1.1']) },
      u2: { _id: 'u2', name: 'jane', locationIds: Immutable.Set(['p2.1', 'p2.2']) },
    });
    checkReducer('loadPlantsSuccess', state, payload, expected);
  });

  test('should delete a plant', () => {
    const state = Immutable.fromJS({
      l1: {
        _id: 'l1',
        name: 'john',
        locationIds: Immutable.Set(['p1.1']),
      },
      l2: {
        _id: 'l2',
        name: 'jane',
        locationIds: Immutable.Set(['p2.1', 'p2.2', 'p2.3']),
      },
    });
    const payload = { locationId: 'l2', plantId: 'p2.1' };
    const expected = Immutable.fromJS({
      l1: { _id: 'l1', name: 'john', locationIds: Immutable.Set(['p1.1']) },
      l2: { _id: 'l2', name: 'jane', locationIds: Immutable.Set(['p2.1', 'p2.2', 'p2.3']) },
    });

    checkReducer('deletePlantRequest', state, payload, expected);
  });
});
