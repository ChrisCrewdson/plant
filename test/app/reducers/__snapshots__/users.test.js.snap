// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/app/reducers/users should delete a plant 1`] = `
Object {
  "l1": Object {
    "_id": "l1",
    "locationIds": Array [
      "p1.1",
    ],
    "name": "john",
  },
  "l2": Object {
    "_id": "l2",
    "locationIds": Array [
      "p2.1",
      "p2.2",
      "p2.3",
    ],
    "name": "jane",
  },
}
`;

exports[`/app/reducers/users should reduce createPlantRequest action 1`] = `
Object {
  "u1": Object {
    "_id": "u1",
    "locationIds": Array [
      "p1",
    ],
    "name": "john",
  },
}
`;

exports[`/app/reducers/users should reduce loadPlantsSuccess action 1`] = `
Object {
  "u1": Object {
    "_id": "u1",
    "locationIds": Array [
      "p1.1",
    ],
    "name": "john",
  },
  "u2": Object {
    "_id": "u2",
    "locationIds": Array [
      "p2.1",
      "p2.2",
    ],
    "name": "jane",
  },
}
`;

exports[`/app/reducers/users should reduce loadUserSuccess action 1`] = `
Object {
  "1": Object {
    "_id": "1",
    "locationIds": Array [],
    "name": "john",
  },
}
`;

exports[`/app/reducers/users should reduce loadUsersSuccess action 1`] = `
Object {
  "1": Object {
    "_id": "1",
    "locationIds": Array [],
    "name": "john",
  },
}
`;

exports[`/app/reducers/users should reduce loadUsersSuccess with undefined payload 1`] = `Object {}`;
